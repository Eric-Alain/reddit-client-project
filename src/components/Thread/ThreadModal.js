//React
import React from 'react'

//3rd party
import PropTypes from "prop-types"
import Img from 'react-cool-img'
import Modal from 'react-bootstrap/Modal'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'

const ThreadModal = ({ show, onHide, img}) => {  
  return (
    <Modal show={show} onHide={onHide} dialogClassName='thread-modal' keyboard={false}>
      <Modal.Header className='justify-content-end mx-3 py-1 px-0'>
        <button className="lh-0 rounded" onClick={onHide}>
          <FontAwesomeIcon className='modal-close' icon={['fas', 'xmark']} size='2x' />
        </button>
      </Modal.Header>
      <Modal.Body className="pt-0">
        <Img className='w-100 rounded' src={img} alt={''} />
      </Modal.Body>
    </Modal>
  )
}

ThreadModal.propTypes = {
  show: PropTypes.bool.isRequired,
  onHide: PropTypes.func.isRequired,
  img: PropTypes.string.isRequired
}

export default ThreadModal