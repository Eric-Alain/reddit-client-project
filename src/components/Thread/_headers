/*
  X-Content-Type-Options: nosniff
  X-Frame-Options: SAMEORIGIN
  X-XSS-Protection: 1; mode=block
    
  Content-Security-Policy: base-uri 'self';
  Content-Security-Policy: connect-src 'self' https://*.cloudfront.net 'unsafe-inline';
  Content-Security-Policy: default-src 'self' 'unsafe-inline';
  Content-Security-Policy: form-action 'self';
  Content-Security-Policy: font-src 'self' data: https://*.cloudfront.net https://fontawesome.com;
  Content-Security-Policy: img-src 'self';
  Content-Security-Policy: media-src 'self';
  Content-Security-Policy: object-src 'none';
  Content-Security-Policy: prefetch-src 'self' https://*.cloudfront.net 'unsafe-inline';
  Content-Security-Policy: script-src 'self' https://*.cloudfront.net 'unsafe-inline';
  Content-Security-Policy: script-src-elem 'self' https://*.cloudfront.net 'unsafe-inline';
  Content-Security-Policy: style-src 'self' 'unsafe-inline';
  