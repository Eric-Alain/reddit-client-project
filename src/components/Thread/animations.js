import { gsap } from 'gsap'
import { ScrollTrigger } from 'gsap/ScrollTrigger'

export const animations = refs => {
  gsap.registerPlugin(ScrollTrigger)

  refs.forEach((el) => {
    gsap.fromTo(
      el.querySelector('.thread-img'),
      {
        autoAlpha: 0
      },
      {
        duration: 1,
        autoAlpha: 1,
        ease: 'none',
        scrollTrigger: {
          trigger: el,
          start: 'top bottom',
          end: 'top center',
          scrub: true,
          toggleActions: 'play none none reverse',
          invalidateOnRefresh: true
        }
      }
    )
  })
}
