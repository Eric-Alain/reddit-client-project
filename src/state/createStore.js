//Redux
import { createStore as reduxCreateStore, combineReducers } from 'redux'
import { composeWithDevTools } from 'redux-devtools-extension'

//Reducers
import searchReducer from '../state/reducers/search'
import dataReducer from '../state/reducers/data'
import subredditsReducer from '../state/reducers/subreddits'
import togglesReducer from '../state/reducers/toggles'

//Create root reducer
export const rootReducer = combineReducers({
  search: searchReducer,
  data: dataReducer,
  subreddits: subredditsReducer,
  toggles: togglesReducer
})

//Create store, pass preloadedState via plugin
const createStore = preloadedState => {
  return reduxCreateStore(rootReducer, preloadedState, composeWithDevTools())
}
export default createStore
