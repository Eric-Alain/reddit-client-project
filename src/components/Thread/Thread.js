//React
import React, { useEffect, useRef } from 'react'

//Redux
import { useDispatch, useSelector } from 'react-redux'
import { fetchData, limitDataResults } from '../../state/actions/data'
import { setModalData, toggleModal, isLoading } from '../../state/actions/toggles'

//Utils
import ThreadModal from '../../components/Thread/ThreadModal'
import { animations } from '../../components/Thread/animations'
import { renderDate, nFormatter, getSmallestImage } from '../../utils/utils'
import { threadFetch } from '../../utils/fetches'

//3rd party
import Img from 'react-cool-img'
import Col from 'react-bootstrap/Col'
import Row from 'react-bootstrap/Row'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'

const Thread = () => {
  //Redux
  const dispatch = useDispatch()
  const data = useSelector(state => state.data.reddit)
  const limit = useSelector(state => state.data.limit)
  const modalData = useSelector(state => state.toggles.modalData)
  
  //Refs
  const refArr = useRef([])
  refArr.current = []

  const addToRefs = el => {
    if (el && !refArr.current.includes(el)) {
      refArr.current.push(el)
    }
  }

  //Handlers
  const setActiveModal = index => {
    dispatch(setModalData({ activeModal: index }))
    dispatch(toggleModal(true))
  }

  const setInactiveModal = () => {
    dispatch(setModalData({ activeModal: null }))
    dispatch(toggleModal(false))
  }

  //UseEffects
  useEffect(() => {
    //Async ensures that we are waiting for our fetch to complete before proceeding
    threadFetch('Web Dev').then(res => {
      dispatch(fetchData(res))
      dispatch(isLoading(false))
    })
    animations(refArr.current)
  }, [dispatch])

  //Intentionally want this to run on every render
  //Want to re-evaluate how many refs need to be animated
  useEffect(() => {
    animations(refArr.current)
  })

  return (
    <>
      <Col xs={{ span: 10, order: 2 }} sm={{ span: 7, order: 1 }} md={{ span: 8, order: 1 }} xl={{ span: 9, order: 1 }} id='thread' className='bg-2'>
        <section>
          <Row>
            <Col xs='12'>
              <h2 className='mt-3 mb-2'>Thread</h2>
              <hr className='my-0' />
            </Col>

            {/*Map data to DOM once available, limit results based on state limit*/}
            {data !== undefined ? (
              data.slice(0, limit ? limit : data.length).map((child, i) => {

                return (
                  <article key={i}>
                    <Col xs='12' className='py-3 mb-3' ref={addToRefs}>
                      <h3>
                        <a href={child.url}>{child.title}</a>
                      </h3>
                      <div className='img-container' onClick={() => setActiveModal(i)} onKeyDown={() => setActiveModal(i)} role='button' aria-label='preview-image' tabIndex='0'>
                        <div className={`img-undefined justify-content-center align-items-center ${!child.previewImage ? 'show' : 'hide'}`}>
                          <p className='img-undefined-text rounded'>Preview image unavailable</p>
                        </div>
                        <Img
                          className='thread-img w-100 rounded'
                          src={child.previewImage !== undefined ? getSmallestImage(child.previewImage) : 'https://placekitten.com/g/480/320'}
                          alt={''}
                          height={280}
                          width={160}
                        />
                      </div>
                      <ThreadModal
                        show={modalData.activeModal === i}
                        onHide={() => setInactiveModal()}
                        img={child.previewImage !== undefined ? getSmallestImage(child.previewImage) : 'https://placekitten.com/g/480/320'}
                      />
                      <hr />
                      <Row className='align-items-start justify-content-start justify-content-lg-between'>
                        <Col xs='12' lg='7' xl='8'>
                          <Row>
                            <Col xs='auto' className='pe-2'>
                              <FontAwesomeIcon icon={['fab', 'reddit-square']} size='1x' />
                            </Col>
                            <Col xs='auto' className='ps-0'>
                              <small>
                                <a href={`https://www.reddit.com/user/${child.author}`}>{child.author}</a> |{' '}
                                <a href={`https://www.reddit.com/${child.subredditNamePrefixed}`}>{child.subredditNamePrefixed}</a>: {nFormatter(child.subredditSubscribers, 1)}
                              </small>
                            </Col>
                          </Row>
                        </Col>
                        <Col xs='12' lg='auto'>
                          <Row className='justify-content-start justify-content-lg-end'>
                            <Col xs='auto' className='pe-2'>
                              <FontAwesomeIcon icon={['far', 'calendar-days']} size='1x' />
                            </Col>
                            <Col xs='auto' className='ps-0'>
                              <small>{renderDate(child.createdUtc)}</small>
                            </Col>
                          </Row>
                        </Col>
                        <Col xs='12' lg='auto'>
                          <Row className='justify-content-start justify-content-lg-end'>
                            <Col xs='auto' className='pe-2'>
                              <FontAwesomeIcon icon={['far', 'message']} size='1x' />
                            </Col>
                            <Col xs='auto' className='ps-0'>
                              <small>
                                <a href={`https://www.reddit.com${child.permalink}`}>{child.numComments}</a>
                              </small>
                            </Col>
                          </Row>
                        </Col>
                      </Row>
                    </Col>
                  </article>
                )
              })
            ) : (
              /*If data unavailable, insert a placeholder element*/
              <article>
                <Col xs='12' className='py-3 mb-1' key={0} ref={addToRefs}>
                  <Col xs='12' className='loading-placeholder loading-pulse rounded'></Col>
                  <div className='img-container'>
                    <Img className='thread-img rounded loading loading-pulse' src={null} alt={null} />
                  </div>
                </Col>
              </article>
            )}

            {/*Load more element*/}
            <Col xs='12' className='text-center py-3'>
              <hr />
              <button
                className='more-btn my-2 p-2 rounded'
                onClick={() => {
                  dispatch(limitDataResults(limit + 5))
                }}
              >
                Show more results...
              </button>
            </Col>
          </Row>
        </section>
      </Col>
    </>
  )
}

export default Thread
