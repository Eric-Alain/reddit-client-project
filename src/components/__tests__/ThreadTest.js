import React from 'react'
import renderer from 'react-test-renderer'
import Thread from '../Thread/Thread'

import { Provider } from 'react-redux'
import createStore from '../../state/createStore'
import registerIcons from '../../icons/icons'

const store = createStore()
registerIcons()

describe('Thread', () => {
  it('renders correctly', () => {
    const tree = renderer
      .create(
        <Provider store={store}>
          <Thread />
        </Provider>
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })
})
