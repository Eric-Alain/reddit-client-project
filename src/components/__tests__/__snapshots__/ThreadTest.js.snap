// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Thread renders correctly 1`] = `
<div
  className="bg-2 col-xl-9 col-md-8 col-sm-7 col-12 order-xl-1 order-md-1 order-sm-1 order-2"
  id="thread"
>
  <section>
    <div
      className="row"
    >
      <div
        className="col-12"
      >
        <h2
          className="mt-3 mb-2"
        >
          Thread
        </h2>
        <hr
          className="my-0"
        />
      </div>
      <article>
        <div
          className="py-3 mb-1 col-12"
        >
          <div
            className="loading-placeholder loading-pulse rounded col-12"
          />
          <div
            className="img-container"
          >
            <img
              alt={null}
              className="thread-img rounded loading loading-pulse no-js-"
              src="data:image/gif;base64,R0lGODlhAQABAIAAAP///wAAACH5BAEAAAAALAAAAAABAAEAAAICRAEAOw=="
            />
            <noscript>
              <style>
                .no-js- { display: none !important; }
              </style>
              <img
                alt={null}
                className="thread-img rounded loading loading-pulse"
                src={null}
              />
            </noscript>
          </div>
        </div>
      </article>
      <div
        className="text-center py-3 col-12"
      >
        <hr />
        <button
          className="more-btn my-2 p-2 rounded"
          onClick={[Function]}
        >
          Show more results...
        </button>
      </div>
    </div>
  </section>
</div>
`;
